system/mp/mpcomm.cpp

```diff
@@ -762,22 +886,42 @@ protected: friend class CMPPacketReader;
                     throw e;
                 }
 #ifdef _TRACE
+                StringBuffer str;
                 str.clear();
                 LOG(MCdebugInfo(100), unknownJob, "MP: Retrying connection to %s, %d attempts left",remoteep.getUrlStr(str).toCharArray(),retrycount+1);
 #endif
             }
+
             ::Release(newsock);
             newsock = NULL;
+
             {
                 CriticalUnblock unblock(connectsect); // to avoid connecting philosopher problem
 #ifdef _FULLTRACE
-                    PROGLOG("MP: before sleep");
+                PROGLOG("MP: before sleep");
 #endif
-                Sleep(2000+getRandom()%3000);
+                // check often if channelsock was created from accept thread
+                Sleep(50);
+                unsigned totalt = 2000 + getRandom() % 3000;
+                unsigned startt = msTick();
+                unsigned deltat = 0;
+                while (deltat < totalt)
+                {
+                    {
+                        CriticalBlock block(connectsect);
+                        if (channelsock)
+                            break;
+                    }
+                    deltat = msTick() - startt;
+                    Sleep(50);
+                }
 #ifdef _FULLTRACE
-                    PROGLOG("MP: after sleep");
+                PROGLOG("MP: after sleep");
 #endif
             }
+
+            CriticalBlock block(connectsect);
```

Comment:
>I changed it slightly but the point was that connectsect must be held while in the connect loop and so we make sure to hold it here before looping around again after releasing it just above.

Still don't understand what this achieves..
connect() is always called within a connectsect block, the CriticalUnblock on line 899 temporarily releases it.. *and* regains it when the block it's in goes out of scope, i.e. here..
This CriticalBlock would momentarily grab the connectsect CS again (re-enter) and immediately release it, since this 'block' would immediately fall out of scope as it loops around.